File: tree/diameter_of_tree.cpp
/*
TITLE   : 트리의 지름
URL     : https://www.acmicpc.net/problem/1967
DATE    : 21.07.13
*/
#include <iostream>
#include <vector>
#include <utility>
#include <string.h>
#define MAX_LEN 10001

using namespace std;

vector<pair<int,int>> graph[MAX_LEN];
bool visit[MAX_LEN];
int max_cost;
int end_point;

void dfs(int curr, int length){
    visit[curr] = true;
    if(length > max_cost){
        max_cost = length;
        end_point = curr;
    } 
    for(int i=0; i<graph[curr].size(); i++){
        int next = graph[curr][i].first;
        if(!visit[next]){
            int cost = graph[curr][i].second;
            dfs(next, length + cost);
        }
    }
}

int main(){
    int n;
    scanf("%d",&n);
    for(int i=0; i<n-1; i++){
        int a, b, c;
        scanf("%d %d %d",&a,&b,&c);
        graph[a].push_back(make_pair(b,c));
        graph[b].push_back(make_pair(a,c));
    }

    //dfs
    dfs(1,0);
    max_cost = 0;
    fill(&visit[0],&visit[MAX_LEN],false);
    dfs(end_point,0);
    printf("%d\n",max_cost);

    return 0;
}

File: DFS/diameter_of_tree.py
'''
TITLE   : 트리의 지름
URL     : https://www.acmicpc.net/problem/1167
DATE    : 21.08.26
'''
import sys
from collections import deque

v = int(sys.stdin.readline())
tree = [[] for _ in range(v+1)]
for _ in range(v):
    inform = list(map(int, sys.stdin.readline().split()))
    vnum = inform[0]
    inform = inform[1:-1]
    num = len(inform) // 2
    for i in range(num):
        b, cost = inform[2*i], inform[2*i+1]
        tree[vnum].append((b,cost))

def dfs(start):
    visit = [False] * (v+1)
    s = []
    s.append((start, 0))
    max_dist = 0
    max_node = 0
    while s:
        curr, dist = s[-1]
        visit[curr] = True
        searched = False
        for next in tree[curr]:
            next_node, cost = next
            if not visit[next_node]:
                if max_dist < dist + cost:
                    max_dist = dist + cost
                    max_node = next_node
                s.append((next_node, dist + cost))
                searched = True
                break
        if not searched:
            s.pop(-1)
    return max_node, max_dist

def bfs(start):
    visit = [False] * (v+1)
    q = deque()
    q.append((start, 0))
    visit[start] = True
    max_dist = 0
    max_node = 0
    while q:
        curr, dist = q.popleft()
        for next in tree[curr]:
            next_node, cost = next
            if not visit[next_node]:
                if max_dist < dist + cost:
                    max_dist = dist + cost
                    max_node = next_node
                visit[next_node] = True
                q.append((next_node, dist + cost))
    return max_node, max_dist

next, dist = bfs(1)
next, dist = bfs(next)
        
print(dist)
